src/java.base/share/classes/sun/security/util/Password.java

```diff
@@ -54,17 +54,28 @@ public static char[] readPassword(InputStream in, boolean isEchoOn)
 
         try {
             // Only use Console if `in` is the initial System.in
-            if (!isEchoOn &&
-                    in == SharedSecrets.getJavaLangAccess().initialSystemIn() &&
-                    ConsoleHolder.consoleIsAvailable()) {
-                consoleEntered = ConsoleHolder.readPassword();
-                // readPassword returns "" if you just press ENTER with the built-in Console,
-                // to be compatible with old Password class, change to null
-                if (consoleEntered == null || consoleEntered.length == 0) {
-                    return null;
+            if (!isEchoOn) {
+                if (in == SharedSecrets.getJavaLangAccess().initialSystemIn()
+                        && ConsoleHolder.consoleIsAvailable()) {
+                    consoleEntered = ConsoleHolder.readPassword();
+                    // readPassword returns "" if you just press ENTER with the built-in Console,
+                    // to be compatible with old Password class, change to null
+                    if (consoleEntered == null || consoleEntered.length == 0) {
+                        return null;
+                    }
+                    consoleBytes = ConsoleHolder.convertToBytes(consoleEntered);
+                    in = new ByteArrayInputStream(consoleBytes);
+                } else if (System.in.available() == 0) {
+                    // This may be running in an IDE Run Window or in JShell,
+                    // which acts like an interactive console and echoes the
+                    // entered password. In this case, print a warning that
+                    // the password might be echoed. If available() is not zero,
+                    // it's more likely the input comes from a pipe, such as
+                    // "echo password |" or "cat password_file |" where input
+                    // will be silently consumed without echoing to the screen.
+                    System.err.print("[Warning: Console not available." +
+                            " The entered password might be echoed.] ");
```

Comment:
s/might/may/

I wonder if "echoed" is generally understood by most users. Should we be more clear and say something like " The password entered may be visible on the screen"?

Also, consider putting "Warning" in all caps to make it more significant. Also consider breaking into 2 lines, so the second part is not at the end of a long line, ex:

WARNING: Console not available
WARNING: The password entered may be visible on the screen

Reply:
This text appears after the prompt (which is printed by the caller) and I deliberately make it short so everything is still on the same line.
```
Enter keystore password:  [Warning: Console not available. The entered password might be echoed.] password
```
Shall I just change "echoed" to "visible"? I'm OK with "WARNING".

Reply:
Last line looks good.